Commonalities/documentation/Working/user-consent.md

```diff
@@ -0,0 +1,396 @@
+  - [Purposes](#purposes)
+    - [Purpose levels](#purpose-levels)
+    - [Purpose relationship with other key concepts](#purpose-relationship-with-other-key-concepts)
+  - [Login authorization requirements and purpose usage](#login-authorization-requirements-and-purpose-usage)
+    - [Using purpose parameter in the authorization request](#using-purpose-parameter-in-the-authorization-request)
+    - [Non-automatic purpose requirements for the Login authorization](#non-automatic-purpose-requirements-for-the-login-authorization)
+  - [Obtaining an access token for non-automatic purposes](#obtaining-an-access-token-for-non-automatic-purposes)
+  - [Consent API and GDPR API](#consent-api-and-gdpr-api)
+  - [Annex: Flow source code](#annex-flow-source-code)
+    - [Delegated authorization jwt-bearer flow](#delegated-authorization-jwt-bearer-flow)
+    - [Consent gathering with Consent API flow](#consent-gathering-with-consent-api-flow)
+
+
+## Introduction
+
+Operator NaaS platform implementing CAMARA should be built following a Privacy By Design approach to fully comply with the spirit and letter of the [GDPR regulation](https://gdpr-info.eu/), which provides a great level of protection for user privacy. This means that an API that process Personal Information needs a user consent. Consents are given by users to legal entities to process personal data under a specific purpose.
+
+This document captures guidelines for Operator NaaS platform to handle user consents to comply with GDPR requirements in an easy way. It introduces the concept of **purpose** in OpenID Connect, which is built on top of OAuth 2.0. These User consent management guidelines define a procedure for application back-ends to obtain access tokens for purposes which require user consent. It relies on an existing previous authorization session of the user (where user identity was validated against the corresponding Identity Provider). Application back-ends will be then able to get access tokens with the required purposes on behalf of the user (i) referring to a previous authorization, and (ii) without requiring further user interaction in between, once the user consent is collected. 
+
+Auth server will not manage the user consent during authorization process. Consent gathering from users is a one-off independent process that must be completed before the actual access to the information.
+
+The above also allows Operator NaaS platform to have consent control centralized in a third party instead of being done in Auth Server. Auth Server will check if the user has provided the consent to the purposes requested before issuing an access token.
+
+## Glossary of Terms and Concepts
+
+The list below introduces several key concepts on how to handle user personal information data.
+
+- `Application`: client system that requires access to protected resources. Application must use the appropriate access token to access those resources.
+- `User`: the resource owner.
```

Comment:
The definition of "user" needs to be expanded so that I understand the full scope of this proposal. We'll get to the definition of "resource" shortly. But first, is the "user" here always a subscriber of the network providing the API (i.e. a mobile subscriber or fixed broadband subscriber)?
- If so, is their role in this process that of the network subscriber, or could it be as a customer of the "app", or both? Who does the "user" have a contractual relationship with in this process?
- If the "user" may not be the network subscriber, who else could the "user" be and what would be their contractual relationship be with the network API provider?

As for "resource", well, this at least needs a definition. However, the definition below for "resource server" suggests that the "resource" is something stored within the network - in other words "resource" is just data and, given that the "user" owns this, then it is personal data. That would fit with the GDPR focus of this proposal.

If it is just personal data, then I'm starting to get confused by "consent" in this context. For an API such as the Device Status API, where the API consumer passes in, for example, MSISDN, and gets back the roaming status for that subscriber, it can certainly be argued that status information is "location data" and thus personal data to which the "resource owner" (i.e. mobile subscriber) would need to consent to being passed to the API consumer.

But in another case, the QoD API, the API consumer would pass in some personal data (again, say the MSISDN) and some other parameters about the flow, and would get back .... well, just confirmation that the enhanced quality for the flow has been activated (or not), but no additional personal data. So that API is not accessing any personal data at all, and hence I am confused as to why the subscriber must give GDPR consent for purpose "improve-network-performance". Are we saying that networks must have "opt-in" GDPR consent before they can make any effort to improve the quality of a subscriber's connection?

Or is the proposal really saying that each API should have defined a set of "purposes" to which the target network subscriber (mobile or fixed) would need to "consent" to before the API could be actioned for that subscriber? So, for the QoD API, it would be defined that "improve-network-performance" consent is required before an API consumer can request enhanced QoS for a given subscriber, even though this is nothing to do with GDPR. Whereas opt-in "retrieve location data" consent will be required before an API consumer can request the roaming status of a subscriber using the Device Status API (in this case, because of GDPR regulations).

In other words, is this a general "user (i.e. subscriber) consent" proposal, of which GDPR is only one reason why consent may be required? If so, then the document should be re-written so that it is not so GDPR centric, and does not assume that the API consumer is always trying to access subscriber personal data (that they do not already know) via the API.

Reply:
>The definition of "user" needs to be expanded so that I understand the full scope of this proposal. We'll get to the definition of "resource" shortly. But first, is the "user" here always a subscriber of the network providing the API (i.e. a mobile subscriber or fixed broadband subscriber)?
If so, is their role in this process that of the network subscriber, or could it be as a customer of the "app", or both? Who does the "user" have a contractual relationship with in this process?
If the "user" may not be the network subscriber, who else could the "user" be and what would be their contractual relationship be with the network API provider?

On what refers to CAMARA APIs to be exposed by Telco operators, the end user is the human participant which is identified in Telco Operator by a unique user identifier (e.g. Subject identifier sub in OIDC terminology). Therefore, the Authentication process allows to confirm/validate user identity.

We are referring as user as the person who uses the Operator capabilities (end user of the device). But normally they will be the person who has a contract relationship with the Telco Operator (the one paying the bills) unless the operator have specific features to identify whether the end user using the device is not the one having the contract relationship with the Operator... which I guess you mean by subscriber.

Just to provide an example, in case of TEF in Brazil, there are family plans with multiple phone numbers. One "owner" phone number (the one corresponding to the user having the contract relationship) and other dependent phone numbers (used by other family members not being the owner). And we have the mechanisms in place to have different users for each phone number (different user id and profile). Otherwise, from Operator perspective the end user will be always the subscriber.

Reply:
>As for "resource", well, this at least needs a definition. However, the definition below for "resource server" suggests that the "resource" is something stored within the network - in other words "resource" is just data and, given that the "user" owns this, then it is personal data. That would fit with the GDPR focus of this proposal.

Yes, in CAMARA context, resources basically refers to APIs exposed by the Telco Operator.  

>If it is just personal data, then I'm starting to get confused by "consent" in this context. For an API such as the Device Status API, where the API consumer passes in, for example, MSISDN, and gets back the roaming status for that subscriber, it can certainly be argued that status information is "location data" and thus personal data to which the "resource owner" (i.e. mobile subscriber) would need to consent to being passed to the API consumer.

>But in another case, the QoD API, the API consumer would pass in some personal data (again, say the MSISDN) and some other parameters about the flow, and would get back .... well, just confirmation that the enhanced quality for the flow has been activated (or not), but no additional personal data. So that API is not accessing any personal data at all, and hence I am confused as to why the subscriber must give GDPR consent for purpose "improve-network-performance". Are we saying that networks must have "opt-in" GDPR consent before they can make any effort to improve the quality of a subscriber's connection?

 According to the feedback we've got so far from Legal teams, QoD API for example, will require user consent. But
IMHO, this is more a legal discussion rather than a technical one. Basically, legally speaking it would be required to validate which CAMARA APIs need a user consent which is is out of the scope of this document. This document just defines a technical solution to handle user consent when it is required.

>Or is the proposal really saying that each API should have defined a set of "purposes" to which the target network subscriber (mobile or fixed) would need to "consent" to before the API could be actioned for that subscriber? So, for the QoD API, it would be defined that "improve-network-performance" consent is required before an API consumer can request enhanced QoS for a given subscriber, even though this is nothing to do with GDPR. Whereas opt-in "retrieve location data" consent will be required before an API consumer can request the roaming status of a subscriber using the Device Status API (in this case, because of GDPR regulations).

The proposal says that it is required to define a "purpose" (e.g. improve-network-performance) which represents the reason why a client whats to access to a specific API. So, the purpose definition maps to some specific scope/s which give access to the required API (e.g. QoD API). And the consent is provided by the user to access QoD API under the purpose "improve-network-performance". And the API client will be required to obtain an access token to invoice QoD API under "improve-network-performance" purpose, but that token will be issued by the Auth Server only if the user consent has been previously registered for that application client and that purpose.   

More than one purpose may map to same API, as users may provide consent to access to that API for a functional motive but not for commercial/advertising motives for example.  

>In other words, is this a general "user (i.e. subscriber) consent" proposal, of which GDPR is only one reason why consent may be required? If so, then the document should be re-written so that it is not so GDPR centric, and does not assume that the API consumer is always trying to access subscriber personal data (that they do not already know) via the API.

This will be improved to avoid misunderstanding. The initial goal of the technical solution was to meet GDPR legal requirements in EU. That's why GDPR is mention and that why some concepts directly maps with GDPR definitions. But actually the solution and concepts are generic enough to manage user consents for non-GDPR countries. It i not GDPR specific.